pypeit/scripts/utils.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
+#
+# See top-level LICENSE file for Copyright information
+#
+# -*- coding: utf-8 -*-
+"""
+This file contains a bunch of utility functions
+that can be used with the PypeIt scripts.
+"""
+
+import os
+import numpy as np
+from configobj import ConfigObj
+
+from astropy.table import Table
+
+from pypeit.spectrographs.util import load_spectrograph
+from pypeit.par import PypeItPar
+from pypeit.par.util import parse_pypeit_file
+from pypeit.metadata import PypeItMetaData
+from pypeit.core import procimg
+from pypeit import msgs
+from pypeit import slittrace
+from pypeit import masterframe
+
+
+class Utilities:
+    def __init__(self, args):
+        self.pypeit_file = args.file
+        self.det = args.det
+        self.iFile = None  # if a single frame is being used, set the index, otherwise None
+        self.spectrograph = None
+        self.par = None
+
+        # Load the pypeit file
+        self.cfg_lines, self.data_files, self.frametype, self.usrdata, self.setups =\
+            parse_pypeit_file(args.file, runtime=False)
+
+    def check_index(self, iFile):
+        if self.iFile is None and iFile is None:
+            msgs.error("file index must be selected to get base name")
+        elif iFile is None:
+            iFile = self.iFile
+        return iFile
+
+    def get_basename(self, iFile=None):
+        idx = self.check_index(iFile)
+        # Grab the filename
+        fname = self.data_files[idx]
+        return os.path.basename(fname)
+
+    def get_master_dirkey(self, iFile=None):
+        # Do some checks
+        if self.spectrograph is None or self.par is None:
+            msgs.error()
+        idx = self.check_index(iFile)
+        # Get the master dir
+        if self.par['calibrations']['master_dir'] == 'default':
+            mdir = os.path.join(self.par['rdx']['redux_path'], 'Masters')
+        else:
+            mdir = self.par['calibrations']['master_dir']
+
+        if not os.path.exists(mdir):
+            mdir_base = os.path.join(os.getcwd(), os.path.basename(mdir))
+            msgs.warn('Master file dir: {0} does not exist. Using {1}'.format(mdir, mdir_base))
+            mdir = mdir_base
+        # Get the base name
+        file_base = self.get_basename(iFile=iFile)
+        ftdict = dict({file_base: 'science'})
+        fitstbl = PypeItMetaData(self.spectrograph, self.par,
```

Comment:
Are you sure there are multiple scripts that need `fitstbl`?

I can only think of the `PypeIt` class and even its script doesn't use it.

Reply:
So, I'm not actually storing the `fitstbl` here, I'm using it to generate the master key and master dir (based only on the input `.pypeit` file). I think this has to be loaded if I'm not passing a Master calibration frame as input?

Reply:
Can't remember myself.  Not sure if @kbwestfall will know..

Reply:
I honestly don't remember either, but it does look like it's needed if you want to generate the master key correctly.